### PR TITLE
Fix Scala Wart about implicit () class parameters

### LINKED ISSUE #2576: Scala Wart: Classes cannot have only implicit parameter lists
Opening this issue, as suggested by Martin, to provide a place to discuss the individual warts brought up in the blog post [Warts of the Scala Programming Language](http://www.lihaoyi.com/post/WartsoftheScalaProgrammingLanguage.html) and the possibility of mitigating/fixing them in Dotty (and perhaps later in Scala 2.x). These are based on Scala 2.x behavior, which I understand Dotty follows closely, apologies in advance if it has already been fixed

--------------------------------------------------------------------------------

This doesn't work:

```scala
@ class Foo(i: Int)
defined class Foo

@ new Foo(1)
res50: Foo = $sess.cmd49$Foo@7230510

@ class Bar(implicit i: Int)
defined class Bar

@ new Bar(1)
cmd52.sc:1: no arguments allowed for nullary constructor Bar: ()(implicit i: Int)$sess.cmd51.Bar
val res52 = new Bar(1)
                    ^
Compilation Failed
```

But this does:

```scala
@ new Bar()(1)
res52: Bar = $sess.cmd51$Bar@467de021
```

This one straddles the line between "Wart" and "Bug", but definitely should
be fixed so that a class defined with one argument list doesn't magically 
sprout two.

### PR DESCRIPTION
**Pull request by odersky**
Fixes #2576

As the discussion in #2576 shows, we still have some problems with the implicitly
inserted empty parameter lists for class constructors. We do need that empty list
to support syntax like `C()` and `new C()`. But it gets in the way if a class has
using clauses. Example from the issue:
```scala
class Bar(using x: Int)(y: String)
given Int = ???
def test = new Bar("")
```
Here, an implicitly inserted `()` in front makes the last line fail. We'd need
`new Bar()("")`.

If a class has only using clauses as parameters we now insert a `()` at the end
instead of at the start. That makes the example compile.

For old-style implicit parameters we don't have a choice. We still need the `()` at
the start since otherwise we'd change the meaning of calls with explicit arguments
for the implicit parameters.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala [hunk 1]
```diff
@@ -1135,7 +1135,23 @@ class TreeUnpickler(reader: TastyReader,
               tpd.Super(qual, mixId, mixTpe.typeSymbol)
             case APPLY =>
               val fn = readTerm()
-              tpd.Apply(fn, until(end)(readTerm()))
+              val args = until(end)(readTerm())
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1138]: I've added a commit to test backwards compatibility and when inlining a def from a previous compiler version, we crash when unpickling the arguments here.

**Review by smarter — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1168]: Ah, I handled only one part before (where all parameters are using clauses). Now the mixed case is handled as well.

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala [hunk 2]
```diff
@@ -1523,4 +1555,14 @@ object TreeUnpickler {
   inline val AllDefs = 2          // add everything
 
   class TreeWithoutOwner extends Exception
+
+  /** An attachment key indicating that an old-style leading () in a constructor
+   *  call that has otherwise only using clauses was suppressed.
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1560]:
```suggestion
   *  call that is followed by a using clause was suppressed.
```

File: compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala [hunk 3]
```diff
@@ -1523,4 +1555,14 @@ object TreeUnpickler {
...
+   */
+  val SuppressedApplyToNone: Property.Key[Unit] = Property.Key()
+
+  /** An attachment key indicating that an trailing () in a constructor
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:1564]:
```suggestion
  /** An attachment key indicating that a trailing () in a constructor
```

**Review by smarter — COMMENTED** [review]
The backwards compat support LGTM, but I think that forwards compat support via -scala-output-release will require more special-casing (new-style constructors need to be treated like old-style constructors when `ctx.scalaRelease <= ScalaRelease.Release3_1`), I don't have time right now to dig into this but perhaps @prolativ could take a look? It could be a good case study for the planned discussion on -scala-output-release.

**Comment by odersky** [comment]
I don't know how to reconcile this with `-scala-output-release` since it is a source breaking change. We can keep the old encoding under `-source 3.0`, but how that should interact with `-scala-output-release`. There are too many options interacting here for my taste.

**Comment by odersky** [comment]
I think this can be merged now?

**Comment by bishabosha** [comment]
so it looks like this has caused some inference issues: https://github.com/lampepfl/dotty/issues/15171

**Comment by WojciechMazur** [comment]
@odersky can you take a look at reproducer snipper from #15647 for tinkoff/phobos failure
```
import scala.quoted.*
class ProductTypeField(using val quotes: Quotes)(val name: String)
def foo(using Quotes): ProductTypeField = ProductTypeField()("")
```
As I understood your comment for this PR, code `ProductTypeField()("")` should work fine, but `ProductTypeField("")` (without empty params for implicit parameters list) should fail. However, the first case fails in 3.2.x and the second one in 3.1.x

**Comment by odersky** [comment]
Yes, it's the other way round.

**Pull request merged**